src/axom/core/Map.hpp

```diff
@@ -0,0 +1,487 @@
+   */   
+  template <typename Key, typename T>
+  struct Pair{
+    axom_map::Node<Key, T> * first;
+    bool second;
+
+    Pair(axom_map::Node<Key, T> * node, bool status) : first(node), second(status) {}
+  };
+ 
+  /*!
+   * \class Bucket
+   *
+   * \brief Bucket is the implementation of a bucket for this Map with chaining. It's a singly-linked
+   * list, and supports search, insertion, and removal.
+   *
+   * \tparam Key the type of the keys stored in each Node in the list.
+   * \tparapm T the type of the values stored in each Node in the list.
+   */ 
+  template <typename Key, typename T>   
+  class Bucket{
+  public:
+  
+    /// \name Bucket Construction
+    ///@{
+    /*!
+     * \brief Default constructor so new Buckets just construct into nothing. 
+     *  Can likely just be removed. 
+     */  
+    Bucket(){}    
+
+    /*!
+     * \brief Basic constructor, creates singly-linked list with provided size. 
+     *
+     * \param [in] len the number of items this Bucket instance needs to be able to store.
+     *
+     * \pre len > 0
+     */  
+    Bucket(int len){
+      m_list = axom::allocate <axom_map::Node<Key, T> > (len);
+      for(int i = 0; i < len-1; i++){
+        m_list[i].next = i+1;
+      }
+      m_list[len-1].next = -1;
+      m_free = 0;
+      m_head = -1;
+      m_size = 0;
+      m_capacity = len;
+      m_end.key = Key{0};
+      m_end.value = T{0};
+      m_end.next = -2;
+    }
+   
+    /*!
+     * \brief If a Bucket instance is created with the default contructor, this
+     *  is used to initialize and allocate all the member variables. 
+     *
+     * \param [in] len the number of items this Bucket instance needs to be able to store.
+     *
+     * \pre len > 0
+     */  
+    void init(int len){
+      m_list = axom::allocate <axom_map::Node<Key, T> > (len);
+      for(int i = 0; i < len-1; i++){
+        m_list[i].next = i+1;
+      }
+      m_list[len-1].next = -1;
+      m_free = 0;
+      m_head = -1; 
+      m_capacity = len;
+      m_size = 0;
+      m_end.key = Key{0};
+      m_end.value = T{0};
+      m_end.next = -2;
+    }
+    /// @}
+    /// \name Bucket Methods
+    /// @{
+    /*!
+     * \brief Inserts a given key-value pair into the linked list if an item with 
+     *  given key does not already exist. Fails if list is full or item with given key
+     *  already exists.
+     *
+     * \param [in] key the key of the key-value pair to be inserted.
+     * \param [in] value the value of the key-valure pair to be inserted.
+     *
+     * \return Returns a Pair object, with the first element being a pointer to the inserted node
+     *  and the second element being boolean value true if successful. If insertion fails,
+     *  the bool is set to False, and the node pointed to is the end node if the list was full,
+     *  or the item occupying the queried slot if an item with the given key already existed.
+     */
+    axom_map::Pair<Key, T> insert_no_update(Key key, T value){
```

Comment:
Efficiency/API issue: We will likely need to account for `T` types that we'd want to move rather than copy